### PR TITLE
Dmuldrew/smart calculate charging

### PR DESCRIPTION
**Pull request by dmuldrew**
[Pull Request doc](https://breakthrough-energy.github.io/docs/user/git_guide.html#d-pull-request)

### Purpose

Moves the daily smart charging optimization into a function similar to the immediate charging. This will allow both strategies to share a common loop over days of year.

Adds a mermaid architecture diagram a goal for the refactor.

### What the code is doing

1. Use modular arithmetic to simplify adjusted load updates
2. Move smart charging optimization into into function `calculate_daily_smart_charging_trips`

### Testing

Automated testing

### Time estimate

~15min

### CODE REVIEW

File: prereise/gather/demanddata/transportation_electrification/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+## Architecture Diagram
+
+``` mermaid
+graph TD
+    A[Vehicle Demand Interface Function] -->|choose vehicle type, vehicle range, model year, charging strategy| B(Load Data)
+    I --> |evaluate whether charging is available| C{calculate charging over \n given trip window}
```
Comment (BainanXia) [inline prereise/gather/demanddata/transportation_electrification/README.md:6]: I believe we are trying to roll into a new line by `\n` here but I don't see this deployed properly in the preview. Not sure what is the right way to do so in mermaid.

Comment (dmuldrew) [reply prereise/gather/demanddata/transportation_electrification/README.md:6]: Here's the diagram in the branch:
https://github.com/Breakthrough-Energy/PreREISE/tree/dmuldrew/smart_calculate_charging/prereise/gather/demanddata/transportation_electrification

Comment (BainanXia) [reply prereise/gather/demanddata/transportation_electrification/README.md:6]: Yeah, this is what I see in the diagram:
image

File: prereise/gather/demanddata/transportation_electrification/daily_trip_charging.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    newdata,
+    input_day,
+    day_iter,
+    data_day,
+    adjusted_load,
+    charging_efficiency,
+    daily_vmt_total,
+    kwh,
+    bev_vmt,
+):
```
Comment (dmuldrew) [inline prereise/gather/demanddata/transportation_electrification/daily_trip_charging.py:22]: I'm still adding doc strings to this function and might remove a couple parameters

File: prereise/gather/demanddata/transportation_electrification/README.md (hunk 1 above)
Comment (dmuldrew) [reply prereise/gather/demanddata/transportation_electrification/README.md:6]: Interesting, it worked in the live online editor:
https://mermaid.live/edit#pako:eNp1kk-L2zAQxb_K4FMWEnrPobAbdzdZylJIaQ_xEmalsS2QJaM_Dm6c796x5WRJoTpZ8k_vzTzNORNWUrbOKodtDT_zwgCvx8MvqpXQBDk1aCTsTCBXoiB4jkYEZc07rFZfB1Fb6wm6mQ59S8vbzqGpeNuwg4ae0C1B1OgqZSrwwWGgqh_gafHdooQcAz4k990oDQN1qCNDcKop1OQ-LysP2KHS-KFpgM1ZoBZRj-gNsR1fKFivUh0ZCE61cFJG2tMFksvr5LI9vJD1LQlVKgFbG53uYSroh7Ol0vSe6O1EPx9KXJe4Eujgn4Rm3N_x3245Jkltwz3wcg_sAwblgxJX6ilF0aLjlG0M4EhzLiaAsDo2xg-wO2yuTc-hKK1CD2NR15RZLen5-DG_NAdyPB5_T5GwuW0TMK7NZJov3qxrUKs_JCd87vThE8xTdZ7zJyitA8n-_ZfGmlBzjNKpbiyrxcDTY_xyGgL-4ajiCUIN482R4MkK1nEzr4v9KPY_w1TZoLnc9MISew_KTMI8CYkkI7Nl1hBXryTP9nk8LjKeoYaKbM2fkkqMOhRZYS6MYgx23xuRrYOLtMxiKzm2XCFn1aTDy1-mpBAr

File: prereise/gather/demanddata/transportation_electrification/smart_charging.py [hunk 3]
```diff
@@ -103,194 +102,24 @@ def smart_charging(
...
-                    # fval is the value of the final cost, exitflag is the reason why the optimization terminates
-                    # 0-success, 1-limit reached, 2-problem infeasible, 3-problem unbounded, 4-numerical difficulties
-                    x = np.array(linprog_result.x)
-                    exitflag = linprog_result.status
-
-                    state_of_charge = np.zeros((total_trips, 2))
-
-                    # find the feasible points
-                    if exitflag == 0:
+        adjusted_load = load_demand + model_year_profile
```
Comment (BainanXia) [inline prereise/gather/demanddata/transportation_electrification/smart_charging.py:113]: Wonder what happened here, did we preprocessed `load_demand` and `model_year_profile` somewhere else? Also, if that's the case should we move it out of the loop given it is not affected by the iterator `day_iter`.

Comment (dmuldrew) [reply prereise/gather/demanddata/transportation_electrification/smart_charging.py:113]: It's affected by the iteration, this line is adding current state of the model year profile to the base load. 

Here's the new simplified daily loop:

Screen Shot 2022-11-29 at 12 32 13 PM

Comment (BainanXia) [reply prereise/gather/demanddata/transportation_electrification/smart_charging.py:105]: I still don't see how `adjusted_load` is changed in the iteration. It seems we are only slicing it with different `trip_window_indices`. Also, in `calculate_daily_smart_charging_trips`, it seems only the size of `adjusted_load` matters?

Comment (BainanXia) [reply prereise/gather/demanddata/transportation_electrification/smart_charging.py:113]: I see. I was wondering how the refactored logic is equivalent to the original implementation.

**Comment by dmuldrew** [comment]
Last commit makes implements the following changes:
https://github.com/Breakthrough-Energy/PreREISE/issues/325
which breaks the existing integration tests.

The current ratio of total yearly BEV demand for immediate and smart charging is:
`immediate_BEV_demand.sum()/smart_BEV_demand.sum() = 1170.520679`

**Comment by dmuldrew** [comment]
Looks like there are definitely optimization failures in smart charging affecting the results:
Screen Shot 2022-12-02 at 3 10 36 PM

**Pull request closed without merging**